README
Mallet-CrfWrapper

The README is used to introduce the module and provide instructions on
how to install the module, any machine dependencies it may have (for
example C compilers and installed libraries) and any other information
that should be provided before the module is installed.

A README file is required for CPAN modules since CPAN extracts the README
file from a module distribution so that people browsing the archive
can use it to get an idea of the module's uses. It is usually a good idea
to provide version information here so that people can decide whether
fixes for the module are worth downloading.


INSTALLATION

To install this module, run the following commands:

	perl Makefile.PL
	make
	make test
	make install

SUPPORT AND DOCUMENTATION

After installing, you can find documentation for this module with the
perldoc command.

    perldoc Mallet::CrfWrapper

You can also look for information at:

    RT, CPAN's request tracker (report bugs here)
        http://rt.cpan.org/NoAuth/Bugs.html?Dist=Mallet-CrfWrapper

    AnnoCPAN, Annotated CPAN documentation
        http://annocpan.org/dist/Mallet-CrfWrapper

    CPAN Ratings
        http://cpanratings.perl.org/d/Mallet-CrfWrapper

    Search CPAN
        http://search.cpan.org/dist/Mallet-CrfWrapper/


LICENSE AND COPYRIGHT

Copyright (C) 2014 David Larochelle

This program is free software; you can redistribute it and/or modify it
under the terms of the the Artistic License (2.0). You may obtain a
copy of the full license at:

L<http://www.perlfoundation.org/artistic_license_2_0>

Any use, modification, and distribution of the Standard or Modified
Versions is governed by this Artistic License. By using, modifying or
distributing the Package, you accept this license. Do not use, modify,
or distribute the Package, if you do not accept this license.

If your Modified Version has been derived from a Modified Version made
by someone other than you, you are nevertheless required to ensure that
your Modified Version complies with the requirements of this license.

This license does not grant you the right to use any trademark, service
mark, tradename, or logo of the Copyright Holder.

This license includes the non-exclusive, worldwide, free-of-charge
patent license to make, have made, use, offer to sell, sell, import and
otherwise transfer the Package with respect to any patent claims
licensable by the Copyright Holder that are necessarily infringed by the
Package. If you institute patent litigation (including a cross-claim or
counterclaim) against any party alleging that the Package constitutes
direct or contributory patent infringement, then this Artistic License
to you shall terminate on the date that such litigation is filed.

Disclaimer of Warranty: THE PACKAGE IS PROVIDED BY THE COPYRIGHT HOLDER
AND CONTRIBUTORS "AS IS' AND WITHOUT ANY EXPRESS OR IMPLIED WARRANTIES.
THE IMPLIED WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR
PURPOSE, OR NON-INFRINGEMENT ARE DISCLAIMED TO THE EXTENT PERMITTED BY
YOUR LOCAL LAW. UNLESS REQUIRED BY LAW, NO COPYRIGHT HOLDER OR
CONTRIBUTOR WILL BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, OR
CONSEQUENTIAL DAMAGES ARISING IN ANY WAY OUT OF THE USE OF THE PACKAGE,
EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.



Changes
Revision history for Mallet-CrfWrapper

0.02    2014-06-03
        * Fix "WebService" method (make it work, remove mentions of Media
          Cloud).
        * Add unit tests for both the "InlineJava" and "WebService" CRF model
          runner methods.
        * "InlineJava": instead of fetching an array with Java objects, fetch
          a single string from Java class and parse it on Perl's side. This is
          needed in order to reduce Inline::Java usage to the minimum and thus
          speed up the model runner by 4x.
        * "WebService": instead of JSON, return a newline-and-space-separated
          plain text as a CRF model runner result. This is how it's being used
          in Inline::Java's case, so we can reuse the same helper subroutines
          and avoid encoding / decoding JSON.
        * Round probabilities to up to 8 digits after decimal point in order to
          make it comparable (thus testable in unit tests).
        * Fixed JUnit unit tests.
        * Some repository cleanup.

0.01    Date/time
        First version, released on an unsuspecting world.



lib/Mallet/java/CrfUtils/src/test/java/org/mediacloud/crfutils/ModelRunnerTest.java
package org.mediacloud.crfutils;

import java.io.FileNotFoundException;
import java.io.IOException;
import java.net.URL;

import junit.framework.TestCase;

import org.apache.commons.io.IOUtils;
import org.apache.commons.lang.StringUtils;

public class ModelRunnerTest extends TestCase {

    private final String test_extractor_model_filename = "crf_extractor_model";
    private final String test_input_txt = "test_input.txt";
    private final String test_output_txt = "test_output.txt";

    private final String extractor_model_path;

    private final String test_input;
    private final String test_output;

    private static String getResourceAsString(String filename) {

        String data;

        try {
            data = IOUtils.toString(ModelRunnerTest.class.getResourceAsStream(filename), "UTF-8");
        } catch (IOException e) {
            System.err.println("File read error: " + e.getMessage());
            data = null;
        }

        return data;
    }

    public ModelRunnerTest() throws IOException {

        URL extractorModelURL = ModelRunnerTest.class.getResource(test_extractor_model_filename);
        if (null == extractorModelURL) {
            throw new IOException("Unable to find test extractor model '" + test_extractor_model_filename + "'");
        }
        this.extractor_model_path = extractorModelURL.getPath();
        if (null == this.extractor_model_path) {
            throw new IOException("Unable to determine extractor model path for text extractor model '" + test_extractor_model_filename + "'");
        }

        this.test_input = getResourceAsString(test_input_txt);
        if (null == this.test_input) {
            throw new IOException("Unable to read test input file '" + test_input_txt + "'");
        }

        this.test_output = getResourceAsString(test_output_txt);
        if (null == this.test_output) {
            throw new IOException("Unable to read test output file '" + test_output_txt + "'");
        }
    }

    public void testCRF() throws IOException, FileNotFoundException, ClassNotFoundException, Exception {

        ModelRunner mr = new ModelRunner(extractor_model_path);

        ModelRunner.CrfOutput[] results = mr.runModelString(this.test_input);
        String resultsString = StringUtils.join(mr.crfOutputsToStrings(results), "\n");

        assertEquals("CRF results are correct", test_output, resultsString);

    }

    public void testCRFReturnString() throws IOException, FileNotFoundException, ClassNotFoundException, Exception {

        ModelRunner mr = new ModelRunner(extractor_model_path);

        String resultsString = mr.runModelStringReturnString(this.test_input);

        assertEquals("CRF results are correct", test_output, resultsString);

    }

    public void testCRFMemory() throws IOException, FileNotFoundException, ClassNotFoundException, Exception {

        ModelRunner mr = new ModelRunner(extractor_model_path);

        String[] results = null;

        for (int x = 0; x < 1000; ++x) {
            results = mr.crfOutputsToStrings(mr.runModelString(this.test_input));
        }

        String resultsString = StringUtils.join(results, "\n");

        assertEquals("CRF results are correct", test_output, resultsString);
    }

}


Makefile.PL
use 5.006;
use strict;
use warnings FATAL => 'all';
use ExtUtils::MakeMaker;

WriteMakefile(
    NAME             => 'Mallet::CrfWrapper',
    AUTHOR           => q{David Larochelle <drlaro@cpan.org>},
    VERSION_FROM     => 'lib/Mallet/CrfWrapper.pm',
    ABSTRACT_FROM    => 'lib/Mallet/CrfWrapper.pm',
    LICENSE          => 'Artistic_2_0',
    PL_FILES         => {},
    MIN_PERL_VERSION => 5.006,
    CONFIGURE_REQUIRES => {
        'ExtUtils::MakeMaker' => 0,
    },
    BUILD_REQUIRES => {
        'Test::More' => 0,
    },
    PREREQ_PM => {
        'Inline::Java' => 0.53,
        'HTTP::Request' => 5.814,
        'HTTP::Status' => 0,
        'LWP::UserAgent' => 0,
        #'ABC'              => 1.6,
        #'Foo::Bar::Module' => 5.0401,
    },
    dist  => { COMPRESS => 'gzip -9f', SUFFIX => 'gz', },
    clean => {
        FILES => 'Mallet-CrfWrapper-* lib/Mallet/java/CrfUtils/target/ _Inline/'
    },
);


lib/Mallet/CrfWrapper_Orig.pm
package Mallet::CrfWrapper;

use 5.006;
use strict;
use warnings FATAL => 'all';

=head1 NAME

Mallet::CrfWrapper - The great new Mallet::CrfWrapper!

=head1 VERSION

Version 0.01

=cut

our $VERSION = '0.01';


=head1 SYNOPSIS

Quick summary of what the module does.

Perhaps a little code snippet.

    use Mallet::CrfWrapper;

    my $foo = Mallet::CrfWrapper->new();
    ...

=head1 EXPORT

A list of functions that can be exported.  You can delete this section
if you don't export anything, such as for a purely object-oriented module.

=head1 SUBROUTINES/METHODS

=head2 function1

=cut

sub function1 {
}

=head2 function2

=cut

sub function2 {
}

=head1 AUTHOR

David Larochelle, C<< <drlaro at cpan.org> >>

=head1 BUGS

Please report any bugs or feature requests to C<bug-mallet-crfwrapper at rt.cpan.org>, or through
the web interface at L<http://rt.cpan.org/NoAuth/ReportBug.html?Queue=Mallet-CrfWrapper>.  I will be notified, and then you'll
automatically be notified of progress on your bug as I make changes.




=head1 SUPPORT

You can find documentation for this module with the perldoc command.

    perldoc Mallet::CrfWrapper


You can also look for information at:

=over 4

=item * RT: CPAN's request tracker (report bugs here)

L<http://rt.cpan.org/NoAuth/Bugs.html?Dist=Mallet-CrfWrapper>

=item * AnnoCPAN: Annotated CPAN documentation

L<http://annocpan.org/dist/Mallet-CrfWrapper>

=item * CPAN Ratings

L<http://cpanratings.perl.org/d/Mallet-CrfWrapper>

=item * Search CPAN

L<http://search.cpan.org/dist/Mallet-CrfWrapper/>

=back


=head1 ACKNOWLEDGEMENTS


=head1 LICENSE AND COPYRIGHT

Copyright 2014 David Larochelle.

This program is free software; you can redistribute it and/or modify it
under the terms of the the Artistic License (2.0). You may obtain a
copy of the full license at:

L<http://www.perlfoundation.org/artistic_license_2_0>

Any use, modification, and distribution of the Standard or Modified
Versions is governed by this Artistic License. By using, modifying or
distributing the Package, you accept this license. Do not use, modify,
or distribute the Package, if you do not accept this license.

If your Modified Version has been derived from a Modified Version made
by someone other than you, you are nevertheless required to ensure that
your Modified Version complies with the requirements of this license.

This license does not grant you the right to use any trademark, service
mark, tradename, or logo of the Copyright Holder.

This license includes the non-exclusive, worldwide, free-of-charge
patent license to make, have made, use, offer to sell, sell, import and
otherwise transfer the Package with respect to any patent claims
licensable by the Copyright Holder that are necessarily infringed by the
Package. If you institute patent litigation (including a cross-claim or
counterclaim) against any party alleging that the Package constitutes
direct or contributory patent infringement, then this Artistic License
to you shall terminate on the date that such litigation is filed.

Disclaimer of Warranty: THE PACKAGE IS PROVIDED BY THE COPYRIGHT HOLDER
AND CONTRIBUTORS "AS IS' AND WITHOUT ANY EXPRESS OR IMPLIED WARRANTIES.
THE IMPLIED WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR
PURPOSE, OR NON-INFRINGEMENT ARE DISCLAIMED TO THE EXTENT PERMITTED BY
YOUR LOCAL LAW. UNLESS REQUIRED BY LAW, NO COPYRIGHT HOLDER OR
CONTRIBUTOR WILL BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, OR
CONSEQUENTIAL DAMAGES ARISING IN ANY WAY OUT OF THE USE OF THE PACKAGE,
EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.


=cut

1; # End of Mallet::CrfWrapper


lib/Mallet/CrfWrapper/InlineJava.pm
#!/usr/bin/perl -w

#
# Package to access CRF model runner via Inline::Java
#
package Mallet::CrfWrapper::InlineJava;

use strict;
use warnings;

use 5.16.2;

use Mallet::CrfWrapper;

use Readonly;
use Data::Dumper;
use Env qw(HOME);
use File::Spec;
use File::Basename;
use File::Slurp;

use Inline (
    Java => 'STUDY',

    # Increase memory available to Java to 1 GB because otherwise the extractor
    # runs out of memory after 1000 extractions
    EXTRA_JAVA_ARGS => '-Xmx1G'
);

my $class_path;

my $pid_connected_to_jvm = $$;

my $use_jni = 0;

my $modelrunner;

my $_crf_source_rt;

BEGIN
{
    use File::Basename;
    use File::Spec;
    use Cwd qw( realpath );

    # lib/CRF/CrfUtils/
    my $file_dir = dirname( __FILE__ );

    $_crf_source_rt = "$file_dir/../";

    $_crf_source_rt = realpath( File::Spec->canonpath( $_crf_source_rt ) );
}

sub _crf_root()
{
    # If this package gets moved to a different location, the subroutine will
    # stop reporting correct paths to MC root, so this is an attempt to warn
    # about the problem early
    if ( __PACKAGE__ ne 'Mallet::CrfWrapper::InlineJava' )
    {
        die 'Package name is not Mallet::CrfWrapper::InlineJava, the package was probably moved to a different location.' .
          ' Please update _crf_root() subroutine accordingly.';
    }

    return $_crf_source_rt;
}

BEGIN
{
    my $jar_dir = _crf_root() . '/CrfWrapper/jars';

    my $jars = [ 'mallet-deps.jar', 'mallet.jar' ];

    #Assumes Unix fix later.
    $class_path = scalar( join ':', ( map { "$jar_dir/$_" } @{ $jars } ) );

    #say STDERR "classpath: $class_path";
}

sub create_model($$$)
{
    my ( $class, $training_data_file, $iterations ) = @_;

    return _create_model_inline_java( $class, $training_data_file, $iterations );
}

sub run_model_inline_java_data_array($$$)
{
    my ( $class, $model_file_name, $test_data_array ) = @_;

    #undef( $modelrunner );

    _reconnect_to_jvm_if_necessary();

    if ( !defined( $modelrunner ) )
    {
        say STDERR "Read model $model_file_name";
        $modelrunner = new org::mediacloud::crfutils::ModelRunner( $model_file_name );
    }

    return _run_model_on_array( $class, $modelrunner, $test_data_array );
}

sub _create_model_inline_java($$$)
{
    my ( $class, $training_data_file, $iterations ) = @_;

    say "Entering _create_model_inline_java()";

    use Inline (
        Java  => 'STUDY',
        STUDY => [
            qw ( cc.mallet.fst.SimpleTagger
              java.io.FileReader java.io.File )
        ],
        AUTOSTUDY => 1,
        JNI       => $use_jni,
        CLASSPATH => $class_path,
        PACKAGE   => 'main'
    );

    _reconnect_to_jvm_if_necessary();

    my $model_file_name = $training_data_file;

    $model_file_name =~ s/\.dat$/Model\.txt/;

    say "Model File: $model_file_name";

    my $foo = cc::mallet::fst::SimpleTagger->main(
        [ "--train", "true", "--iterations", $iterations, "--model-file", $model_file_name, $training_data_file ] );

    return;
}

# reconnect to the JVM if the PID of this process changes
# I'm not sure this is necessary since we're not running in shared JVM mode but better safe -- DRL
sub _reconnect_to_jvm_if_necessary()
{
    if ( $pid_connected_to_jvm != $$ )
    {
        say STDERR "reconnecting to JVM: expected pid $pid_connected_to_jvm actual pid $$";
        Inline::Java->reconnect_JVM();
        $pid_connected_to_jvm = $$;
    }
}

sub _run_model_on_array($$$)
{
    my ( $class, $modelrunner, $test_data_array ) = @_;

    _reconnect_to_jvm_if_necessary();

    my $test_data = join "\n", @{ $test_data_array };

    # If test data is empty
    unless ( $test_data )
    {
        return [];
    }

    # Returning and using a single string from a Java method is way faster than
    # returning and using an array of strings
    my $crf_results_string = $modelrunner->runModelStringReturnString( $test_data );
    unless ( defined $crf_results_string ) {
        die "CRF results are undef.";
    }

    my $results = Mallet::CrfWrapper::convert_crf_string_to_arrayref( $crf_results_string );
    unless ( defined $results ) {
        die "CRF results string is undef.";
    }

    return $results;
}

sub _crf_modelrunner_java_src()
{
    # Compile and prepare Java class from /java/CrfUtils/
    my Readonly $crf_modelrunner_java_path =
      _crf_root() . '/java/CrfUtils/src/main/java/org/mediacloud/crfutils/ModelRunner.java';
    my $crf_modelrunner_java_src = read_file( $crf_modelrunner_java_path );
    unless ( $crf_modelrunner_java_src )
    {
        die "Unable to read the ModelRunner Java class from $crf_modelrunner_java_path.";
    }

    return $crf_modelrunner_java_src;
}

use Inline
  Java      => _crf_modelrunner_java_src(),
  AUTOSTUDY => 1,
  CLASSPATH => $class_path,
  JNI       => $use_jni,
  PACKAGE   => 'main';

1;


lib/Mallet/CrfWrapper/WebService.pm
package Mallet::CrfWrapper::WebService;

#
# Package to access CRF model runner via web service
#

use strict;
use warnings;

use Mallet::CrfWrapper;

use LWP::UserAgent;
use HTTP::Request;
use HTTP::Status qw(:constants);
use Encode;
use URI;

use Data::Dumper;

use constant DEFAULT_CRF_PORT => 8441;

#
# Helpers
#

sub _fatal_error($)
{
    my $error_message = shift;
    Mallet::CrfWrapper::_fatal_error( $error_message );
}

# Return true if the response's error was generated by LWP itself and not by
# the server
sub _response_error_is_client_side($)
{
    my $response = shift;

    if ( $response->is_success )
    {
        die "Response was successful, but I have expected an error.\n";
    }

    my $header_client_warning = $response->header( 'Client-Warning' );
    if ( defined $header_client_warning and $header_client_warning =~ /Internal response/ )
    {
        # Error was generated by LWP::UserAgent; likely we didn't reach server
        # at all (timeout, unresponsive host, etc.)
        #
        # http://search.cpan.org/~gaas/libwww-perl-6.05/lib/LWP/UserAgent.pm#$ua->get(_$url_)
        return 1;
    }
    else
    {
        return 0;
    }
}

my $_crf_server_url = 'http://127.0.0.1:8441/crf';

sub set_webservice_url($)
{
    my $crf_server_url = shift;

    my $uri;
    eval { $uri = URI->new( $crf_server_url )->canonical; };
    if ( $@ )
    {
        _fatal_error( "Invalid CRF model runner web service URI: $crf_server_url" );
    }

    # If someone forgot to explicitly set the port
    my $default_protocol_port = 80;
    if ( $uri->port == $default_protocol_port and ( $crf_server_url !~ /:$default_protocol_port/ ) )
    {
        warn( "CRF model runner web service URL's port was not set, to I'm setting it to " . DEFAULT_CRF_PORT );
        $uri->port( DEFAULT_CRF_PORT );
    }

    $_crf_server_url = $uri->as_string;

    say STDERR "CRF model runner web service URL: $_crf_server_url";
}

#
# Mallet::CrfWrapper "implementation"
#

sub create_model($$$)
{
    my ( $class, $training_data_file, $iterations ) = @_;

    # Clients should use Mallet::CrfWrapper::InlineJava directly
    _fatal_error( "Not implemented in " . __PACKAGE__ );
}

sub run_model_inline_java_data_array($$$)
{
    my ( $class, $model_file_name, $test_data_array ) = @_;

    my $test_data = join "\n", @{ $test_data_array };

    # If test data is empty
    unless ( $test_data )
    {
        return [];
    }

    my $test_data_encoded;
    eval {
        # Have to encode because HTTP::Request only accepts bytes as POST data
        $test_data_encoded = Encode::encode_utf8( $test_data );
    };
    if ( $@ )
    {
        _fatal_error( "Unable to encode_utf8() data: $test_data" );
    }

    # Make a request
    my $ua = LWP::UserAgent->new;
    $ua->timeout( 60 );
    $ua->max_size( undef );

    unless ( $_crf_server_url )
    {
        _fatal_error( "Unable to determine CRF model runner web service URL to use." );
    }

    my $request = HTTP::Request->new( POST => $_crf_server_url );
    $request->content_type( 'text/plain; charset=utf8' );
    $request->content( $test_data_encoded );

    my $response = $ua->request( $request );

    my $results_string;
    if ( $response->is_success )
    {
        # OK
        $results_string = $response->decoded_content;
    }
    else
    {
        # Error; determine whether we should be blamed for making a malformed
        # request, or is it an extraction error

        if ( _response_error_is_client_side( $response ) )
        {
            # Error was generated by LWP::UserAgent; likely we didn't reach server
            # at all (timeout, unresponsive host, etc.)
            _fatal_error( 'LWP error: ' . $response->status_line . ': ' . $response->decoded_content );

        }
        else
        {
            # Error was generated by server

            my $http_status_code = $response->code;

            if ( $http_status_code == HTTP_METHOD_NOT_ALLOWED or $http_status_code == HTTP_BAD_REQUEST )
            {
                # Not POST, empty POST
                _fatal_error( $response->status_line . ': ' . $response->decoded_content );

            }
            elsif ( $http_status_code == HTTP_INTERNAL_SERVER_ERROR )
            {
                # CRF processing error -- die() so that the error gets caught and logged into a database
                die( 'CRF web service was unable to process the download: ' . $response->decoded_content );

            }
            else
            {
                # Shutdown the extractor on unconfigured responses
                _fatal_error( 'Unknown HTTP response: ' . $response->status_line . ': ' . $response->decoded_content );
            }
        }
    }

    unless ( $results_string )
    {
        _fatal_error( "Server returned nothing for POST data: " . $test_data );
    }

    my $results = Mallet::CrfWrapper::convert_crf_string_to_arrayref( $results_string );
    unless ( defined $results ) {
        _fatal_error( "CRF results string is undef." );
    }

    return $results;
}

1;


lib/Mallet/java/CrfUtils/src/main/java/org/mediacloud/crfutils/ModelRunner.java
package org.mediacloud.crfutils;

import cc.mallet.fst.CRF;
import cc.mallet.fst.SumLattice;
import cc.mallet.fst.SumLatticeDefault;
import cc.mallet.pipe.Pipe;
import cc.mallet.pipe.iterator.LineGroupIterator;
import cc.mallet.types.InstanceList;
import cc.mallet.types.Sequence;
import java.io.File;
import java.io.FileInputStream;
import java.io.FileNotFoundException;
import java.io.FileReader;
import java.io.IOException;
import java.io.ObjectInputStream;
import java.io.Reader;
import java.io.StringReader;
import java.text.DecimalFormat;
import java.text.DecimalFormatSymbols;
import java.text.NumberFormat;
import java.util.ArrayList;
import java.util.Map;
import java.util.TreeMap;
import java.util.regex.Pattern;

public class ModelRunner {

    // Format to use for converting Double probability to string
    private final static String PROBABILITY_DOUBLE_FORMAT = "#0.########";

    // CRF instance
    private final CRF crf;

    public ModelRunner(String modelFileName) throws IOException,
            FileNotFoundException, ClassNotFoundException {

        ObjectInputStream s = new ObjectInputStream(new FileInputStream(
                modelFileName));
        crf = (CRF) s.readObject();
        s.close();

        // Mallet's thread safety trick -- http://comments.gmane.org/gmane.comp.ai.mallet.devel/271
        Pipe p = crf.getInputPipe();
        p.getDataAlphabet().stopGrowth();
    }

    private CrfOutput[] runModel(String testFileName) throws Exception {

        InstanceList testData = readTestData(testFileName);
        return runCrfModel(testData);
    }

    public CrfOutput[] runModelString(String testDataString) throws Exception {

        InstanceList testData = readTestDataFromString(testDataString);
        return runCrfModel(testData);
    }

    public String runModelStringReturnString(String testDataString) throws Exception {

        String[] results = crfOutputsToStrings(runModelString(testDataString));
        return joinArrayToString("\n", results);
    }

    private CrfOutput[] runCrfModel(InstanceList testData) {

        /*
         Runtime rt = Runtime.getRuntime();

         System.err.println("Used Memory: " + (rt.totalMemory() - rt.freeMemory()) / 1024 + " KB");
         System.err.println("Free Memory: " + rt.freeMemory() / 1024 + " KB");
         System.err.println("Total Memory: " + rt.totalMemory() / 1024 + " KB");
         System.err.println("Max Memory: " + rt.maxMemory() / 1024 + " KB");
         System.err.println("");
         */
        if (testData.size() > 1) {
            throw new IllegalArgumentException("test data may only contain one sequence");
        }

        Sequence input = (Sequence) testData.get(0).getData();

        return predictSequence(input);

        //return crfOutputsToStrings(crfResults);
    }

    public String[] crfOutputsToStrings(CrfOutput[] crfResults) {
        ArrayList<String> sequenceResults = new ArrayList<String>();

        DecimalFormatSymbols periodAsDecimalSeparator = new DecimalFormatSymbols();
        periodAsDecimalSeparator.setDecimalSeparator('.');

        NumberFormat formatter = new DecimalFormat(PROBABILITY_DOUBLE_FORMAT, periodAsDecimalSeparator);

        for (CrfOutput crfResult : crfResults) {

            StringBuilder line = new StringBuilder();

            line.append(crfResult.prediction);

            // Add (sorted) probabilities
            for (Map.Entry<String, Double> probability : crfResult.probabilities.entrySet()) {
                String key = probability.getKey();
                Double value = probability.getValue();

                line.append(" ");
                line.append(key);
                line.append("=");
                line.append(formatter.format(value));
            }

            sequenceResults.add(line.toString());
        }

        return sequenceResults.toArray(new String[0]);
    }

    private InstanceList readTestData(String testFileName) throws FileNotFoundException {

        Reader testFile = new FileReader(new File(testFileName));
        return instanceListFromReader(testFile);
    }

    private InstanceList readTestDataFromString(final String testData) {

        Reader testFile = new StringReader(testData);
        return instanceListFromReader(testFile);
    }

    private InstanceList instanceListFromReader(Reader testFile) {

        Pipe p = crf.getInputPipe();
        p.setTargetProcessing(false);
        InstanceList testData = new InstanceList(p);
        testData.addThruPipe(
                new LineGroupIterator(testFile,
                        Pattern.compile("^\\s*$"), true));
        return testData;
    }

    public class CrfOutput {

        public String prediction;
        public TreeMap<String, Double> probabilities;
    };

    private CrfOutput[] predictSequence(Sequence input) {

        // That's how SimpleTagger.apply() implements it
        Sequence output = crf.transduce(input);

        try {
            if (output.size() != input.size()) {
                throw new RuntimeException("Failed to decode input sequence " + input);
            }
        } catch (RuntimeException e) {
            System.err.println("Exception: " + e.getMessage());
            return new CrfOutput[0];
        }

        SumLattice lattice = new SumLatticeDefault(crf, input);

        ArrayList<CrfOutput> crfResults = new ArrayList<CrfOutput>();
        for (int j = 0; j < input.size(); j++) {

            //System.err.println(" Input Pos " + j);
            CrfOutput crfResult = new CrfOutput();

            crfResult.probabilities = new TreeMap<String, Double>();

            for (int si = 0; si < crf.numStates(); si++) {
                // to state sj at input position ip
                // double twoStateMarginal = lattice.getXiProbability(j,crf.getState(si),crf.getState(sj));
                // probability of being in state si at input position ip
                double oneStateMarginal = lattice.getGammaProbability(j + 1, crf.getState(si));

                String stateName = crf.getState(si).getName();
                //System.err.println( "Marginal prob: " + stateName + " " +oneStateMarginal );
                crfResult.probabilities.put(stateName, oneStateMarginal);
            }

            String prediction = output.get(j).toString();

            crfResult.prediction = prediction;

            //System.err.println( "Prediction: " + prediction);
            //sequenceResults.add(prediction + " ");
            crfResults.add(crfResult);

        }

        return crfResults.toArray(new CrfOutput[0]);
    }

    private static String joinArrayToString(String glue, String[] array) {

        int arrayLength = array.length;
        if (arrayLength == 0) {
            return null;
        }

        StringBuilder out = new StringBuilder();
        out.append(array[0]);
        for (int x = 1; x < arrayLength; ++x) {
            out.append(glue).append(array[x]);
        }

        return out.toString();
    }
}


lib/Mallet/CrfWrapper.pm
package Mallet::CrfWrapper;

use 5.14.0;
use strict;
use warnings;

=head1 NAME

Mallet::CrfWrapper - a convenient wrapper around Mallet's CRF package.

=head1 VERSION

Version 0.02

=cut

our $VERSION = '0.02';


=head1 SYNOPSIS

This module provides a convenient wrapper around Mallet's CRF package.

It can use either Inline::Java or a WebService to interface with Mallet.

=head1 EXPORT

A list of functions that can be exported.  You can delete this section
if you don't export anything, such as for a purely object-oriented module.

=head1 SUBROUTINES/METHODS

=head2 function1

=cut

#
# Facade to either Inline::Java based or a web service-based CRF model runner
#


# Name of a loaded and active CRF module, either 'Mallet::CrfWrapper::InlineJava' or
# 'Mallet::CrfWrapper::WebService'.
#
# Loading of the module is postponed because Mallet::CrfWrapper::InlineJava compiles
# a Java class and loads it into a JVM in BEGIN{}, which slows down scripts
# that don't have anything to do with extraction
my $_active_crf_module = undef;

my $_webservice_enabled = undef;

my $_webservice_url = undef;

sub use_webservice
{
    my ( $flag ) = @_;

    if ( defined( $flag ) && $flag )
    {
        $_webservice_enabled = 1;
    }
    else
    {
        $_webservice_enabled = 0;
    }
}

sub set_webservice_url
{
    my ( $url ) = @_;

    $_webservice_url = $url;
}

sub _load_and_return_crf_module()
{
    unless ( $_active_crf_module )
    {
        my $module;

        if ( $_webservice_enabled )
        {
            $module = 'Mallet::CrfWrapper::WebService';
        }
        else
        {
            $module = 'Mallet::CrfWrapper::InlineJava';
        }

        eval {
            ( my $file = $module ) =~ s|::|/|g;
            require $file . '.pm';
            $module->import();

            if ( ( $module eq 'Mallet::CrfWrapper::WebService' ) && ( defined( $_webservice_url ) ) )
            {
                $module->set_webservice_url( $_webservice_url );
            }
            1;
        } or do
        {
            my $error = $@;
            _fatal_error( "Unable to load $module: $error" );
        };

        $_active_crf_module = $module;
    }

    return $_active_crf_module;
}

sub create_model($$)
{
    my ( $training_data_file, $iterations ) = @_;

    my $module = _load_and_return_crf_module();

    return $module->create_model( $training_data_file, $iterations );
}

sub run_model_inline_java_data_array($$)
{
    my ( $model_file_name, $test_data_array ) = @_;

    my $module = _load_and_return_crf_module();

    return $module->run_model_inline_java_data_array( $model_file_name, $test_data_array );
}

# Helper to convert string:
#
#     excluded O=0.00008472 excluded=0.54451586 optional=0.45517693 required=0.00022249
#     excluded O=0.00000374 excluded=0.95802485 optional=0.04084096 required=0.00113045
#     ...
#
# to arrayref:
#
# [
#   {
#     'prediction' => 'excluded',
#     'probabilities' => {
#       'O' => '8.472e-05',
#       'required' => '0.00022249',
#       'excluded' => '0.54451586',
#       'optional' => '0.45517693'
#     }
#   },
#   {
#     'prediction' => 'excluded',
#     'probabilities' => {
#       'O' => '3.74e-06',
#       'required' => '0.00113045',
#       'excluded' => '0.95802485',
#       'optional' => '0.04084096'
#     }
#   },
#   ...
# ]
#
sub convert_crf_string_to_arrayref($)
{
    my $string = shift;

    # Format:
    #

    my @result;
    my @lines = split("\n", $string);

    foreach my $line ( @lines ) {

        my @line_parts = split(' ', $line);

        my $prediction = $line_parts[0];
        my $probabilities = {};

        for (my $x = 1; $x <= $#line_parts; ++$x) {
            my ( $name, $value ) = split('=', $line_parts[$x]);
            $probabilities->{ $name } = $value + 0;
        }

        push( @result, { 'prediction' => $prediction, 'probabilities' => $probabilities } );
    }

    return \@result;
}

# Helper
sub _fatal_error($)
{
    # There are errors that cannot be classified as extractor errors (that
    # would get logged into the database). For example, if the whole CRF model
    # runner web service is down, no extractions of any kind can happen anyway,
    # so it's not worthwhile to write a gazillion "extractor error: CRF web
    # service is down" errors to the database.
    #
    # Instead, we go the radical way of killing the whole extractor process. It
    # is more likely that someone will notice that the CRF model runner web
    # service is malfunctioning if the extractor gets shut down.
    #
    # Usual die() wouldn't work here because it is (might be) wrapped into an
    # eval{}.

    my $error_message = shift;

    say STDERR $error_message;
    exit 1;
}


=head1 AUTHOR

David Larochelle, C<< <drlaro at cpan.org> >>

=head1 BUGS

Please report any bugs or feature requests to C<bug-mallet-crfwrapper at rt.cpan.org>, or through
the web interface at L<http://rt.cpan.org/NoAuth/ReportBug.html?Queue=Mallet-CrfWrapper>.  I will be notified, and then you'll
automatically be notified of progress on your bug as I make changes.




=head1 SUPPORT

You can find documentation for this module with the perldoc command.

    perldoc Mallet::CrfWrapper


You can also look for information at:

=over 4

=item * RT: CPAN's request tracker (report bugs here)

L<http://rt.cpan.org/NoAuth/Bugs.html?Dist=Mallet-CrfWrapper>

=item * AnnoCPAN: Annotated CPAN documentation

L<http://annocpan.org/dist/Mallet-CrfWrapper>

=item * CPAN Ratings

L<http://cpanratings.perl.org/d/Mallet-CrfWrapper>

=item * Search CPAN

L<http://search.cpan.org/dist/Mallet-CrfWrapper/>

=back


=head1 ACKNOWLEDGEMENTS

Linas Valiukas made numerous performance improvements and added the WebService code.

=head1 LICENSE AND COPYRIGHT

Copyright 2014 Berkman Center for Internet & Society at Harvard University.

This distribution includes JAR files for Mallet which are licensed under the Common Public License and are Copyright UMass Umherst.

=cut

1; # End of Mallet::CrfWrapper

